### PR TITLE
Stochastic Gradient Descent with Restarts and Snapshot ensembles

### PR DESCRIPTION
**Pull request by zoq**
Implementation of the Stochastic Gradient Descent with Restarts (SGDR) as described in: "SGDR: Stochastic Gradient Descent with Warm Restarts" by I. Loshchilov et al and the Snapshot ensembles technique described in: "Snapshot ensembles: Train 1, get m for free" by G. Huang et al. The implementation reuses the mini batch SGD optimizer class and the main idea is implemented as update and decay policy.

### CODE REVIEW

File: src/mlpack/core/optimizers/sgdr/cyclical_decay.hpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ *
+ * For more information, please refer to:
+ *
+ * @code
+ * @article{Loshchilov2016,
+ *   title   = {Learning representations by back-propagating errors},
+ *   author  = {Ilya Loshchilov and Frank Hutter},
+ *   title   = {{SGDR:} Stochastic Gradient Descent with Restarts},
+ *   journal = {CoRR},
+ *   year    = {2016}
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/sgdr/cyclical_decay.hpp]: Do you think we should add a URL here because this is on arXiv?

File: src/mlpack/core/optimizers/sgdr/sgdr.hpp [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+ * }
+ * @endcode
+ *
+ * @code
+ * @inproceedings{Huang2017,
+ *   title     = {Snapshot ensembles: Train 1, get m for free},
+ *   author    = {Gao Huang, Yixuan Li, Geoff Pleiss, Zhuang Liu,
+ *                John E. Hopcroft, and Kilian Q. Weinberger},
+ *   booktitle = {Proceedings of the International Conference on Learning
+ *                Representations (ICLR)},
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/sgdr/sgdr.hpp]: ICLR proceedings are just on arXiv, I think?  In that case it may also be useful to add a URL here.

File: src/mlpack/core/optimizers/sgdr/sgdr.hpp [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+ * @tparam UpdatePolicyType Update policy used during the iterative update
+ *         process. By default the vanilla update policy
+ *         (see mlpack::optimization::VanillaUpdate) is used.
+ * @tparam DecayPolicyType Decay policy used during the iterative update
+ *         process to adjust the step size (CyclicalDecay or SnapshotEnsembles).
+ */
+template<
+    typename DecomposableFunctionType,
+    typename UpdatePolicyType = MomentumUpdate,
+    typename DecayPolicyType = CyclicalDecay
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/sgdr/sgdr.hpp]: It looks like `DecayPolicyType` can only be `SnapshotEnsembles` or `CyclicalDecay`, and the implementation is different in either case.  So do you think it would be better to simply have two different classes?  I guess that might obviate the need for `DecayPolicyType` in general too; not sure.

File: src/mlpack/core/optimizers/sgdr/sgdr_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  }
+
+  // If a user changed the batch size we have to update the restart fraction
+  // of the cyclical decay instantiation.
+  if (optimizer.BatchSize() != batchSize)
+  {
+    batchSize = optimizer.BatchSize();
+    optimizer.DecayPolicy().EpochBatches() = function.NumFunctions() /
+        double(batchSize);
+  }
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/sgdr/sgdr_impl.hpp]: One possibility here might be to add a function to `DecayPolicy()` called something like `UpdateParameters(OptimizerType&)`, which for `CyclicalDecay` or `SnapshotEnsembles` could do the same thing as this code and for `NoDecay` could do nothing.  This could help reduce the need for two separate implementations here.

File: src/mlpack/core/optimizers/sgdr/sgdr_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    {
+      iterate += optimizer.DecayPolicy().Snapshots()[i];
+    }
+    iterate /= (optimizer.DecayPolicy().Snapshots().size() + 1);
+
+    // Calculate final objective.
+    overallObjective = 0;
+    for (size_t i = 0; i < function.NumFunctions(); ++i)
+      overallObjective += function.Evaluate(iterate, i);
+  }
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/sgdr/sgdr_impl.hpp]: But this part I am less sure of how to move into `SnapshotEnsembles`.  The only thing I could think of is to make a template specialization for `SGD<..., SnapshotEnsembles>` and `MiniBatchSGD<..., SnapshotEnsembles>` but that would be ugly and weird because we can't do partial template member specialization.

File: src/mlpack/core/optimizers/sgdr/cyclical_decay.hpp (hunk 1 above)
Comment (zoq) [inline src/mlpack/core/optimizers/sgdr/cyclical_decay.hpp]: Good idea, I thought I already did this but looks like I didn't.

File: src/mlpack/core/optimizers/sgdr/sgdr.hpp (hunk 3 above)
Comment (zoq) [inline src/mlpack/core/optimizers/sgdr/sgdr.hpp]: I agree, two separate classes one for ``SGDR`` and another one for ``SnapshotSGDR`` is the cleanest way.
